src/compositor/windowpixmapitem.cpp

```diff
@@ -480,13 +496,119 @@ QSGNode *WindowPixmapItem::updatePaintNode(QSGNode *oldNode, UpdatePaintNodeData
     SurfaceNode *node = static_cast<SurfaceNode *>(oldNode);
 
     if (m_item == 0) {
+        node->setTextureProvider(0, false);
         delete node;
         return 0;
     }
 
     if (!node) node = new SurfaceNode;
 
-    node->setTextureProvider(m_item->textureProvider());
+    class SnapshotTextureProvider : public QSGTextureProvider
+    {
+    public:
+        ~SnapshotTextureProvider()
+        {
+            delete fbo;
+            delete program;
+            delete t;
+        }
+        QSGTexture *texture() const Q_DECL_OVERRIDE
+        {
+            return t;
+        }
+        QSGTexture *t;
+        QOpenGLFramebufferObject *fbo;
+        QOpenGLShaderProgram *program;
+        int vertexLocation;
+        int sizeLocation;
+        int textureLocation;
+        int matrixLocation;
+    };
+
+    QSGTextureProvider *provider = m_item->textureProvider();
+    QSGTexture *texture = provider->texture();
+
+    if (!m_hasBuffer && texture) {
+        if (!m_textureProvider) {
+            SnapshotTextureProvider *prov = new SnapshotTextureProvider;
+            m_textureProvider = prov;
+
+            prov->fbo = 0;
+            prov->program = new QOpenGLShaderProgram;
+            prov->program->addShaderFromSourceCode(QOpenGLShader::Vertex,
+                "attribute highp vec4 vertex;\n"
+                "uniform highp mat4 matrix;\n"
+                "uniform highp vec2 size;\n"
```

Comment:
This means nothing for performance, but it reduces the amount of code a tiny bit :)